
# HG changeset patch
# User Alessandro Castellani <alessandro@thunderbird.net>
# Date 1785859592 0
# Node ID f7e855accc21b9322bd66026c3c37fd811598035
# Parent  fd9b4e497f1f7fd7bd775a002a3dad1148ad9583
Bug 1616753 - Fix Gloda filtering when using the filterNode dropdown select. r=tobyp

- Use the Gobal Search to search something.
- Click on the "Attachments" filter.
- Change the select dropdown to something else.
- CLick on the "Attachments" fiulter again.
- Ensure the filters are cleared the results page returns to the original state.

Gloda doesn't have any tests and this section will be deprecated after the Panorama database.
So this is a quick fix to remove a small annoiance that might affect users until next ESR.

Differential Revision: https://phabricator.services.mozilla.com/D315749


diff --git a/mail/base/content/widgets/glodaFacet.js b/mail/base/content/widgets/glodaFacet.js
--- a/mail/base/content/widgets/glodaFacet.js
+++ b/mail/base/content/widgets/glodaFacet.js
@@ -436,19 +436,17 @@
           if (groupPair[0] != null) {
             count += groupPair[1].length;
           }
         }
         this.countNode.textContent = count.toLocaleString();
         return;
       }
 
-      while (this.filterNode.hasChildNodes()) {
-        this.filterNode.lastChild.remove();
-      }
+      this.filterNode.replaceChildren();
 
       const allNode = document.createElement("option");
       allNode.textContent = glodaFacetStrings.GetStringFromName(
         "glodaFacetView.facets.filter." +
           this.attrDef.attributeName +
           ".allLabel"
       );
       allNode.setAttribute("value", "all");
@@ -496,44 +494,51 @@
       this.disabled = !this.trueValues.length;
 
       this.countNode.textContent = this.trueValues.length.toLocaleString();
     }
 
     checkStateChanged() {
       // if they un-check us, revert our value to all.
       if (!this.checked) {
+        this.filterNode.selectedIndex = 0;
+        this.filterNode.value = "all";
         this.selectedValue = "all";
+        this.#triggerFacet(this.trueGroups);
       }
     }
 
     filterChanged() {
       if (!this.checked) {
         return;
       }
       if (this.filterNode.value == "all") {
         this.selectedValue = "all";
-        FacetContext.addFacetConstraint(
-          this.faceter,
-          true,
-          this.trueGroups,
-          false,
-          true
-        );
-      } else {
-        const groupValue = this.realTrueGroups[parseInt(this.filterNode.value)];
-        this.selectedValue = groupValue.category;
-        FacetContext.addFacetConstraint(
-          this.faceter,
-          true,
-          [groupValue],
-          false,
-          true
-        );
+        this.#triggerFacet(this.trueGroups);
+        return;
       }
+
+      const groupValue = this.realTrueGroups[parseInt(this.filterNode.value)];
+      this.selectedValue = groupValue.category;
+      this.#triggerFacet([groupValue]);
+    }
+
+    /**
+     * Trigger the refresh of the Gloda facet view with updated filters.
+     *
+     * @param {Array} groupValue - Array of boolean explicit filtering.
+     */
+    #triggerFacet(groupValue) {
+      FacetContext.addFacetConstraint(
+        this.faceter,
+        true,
+        groupValue,
+        false,
+        true
+      );
     }
   }
 
   customElements.define("facet-boolean-filtered", MozFacetBooleanFiltered);
 
   class MozFacetDiscrete extends HTMLElement {
     constructor() {
       super();
diff --git a/mailnews/db/gloda/modules/Gloda.sys.mjs b/mailnews/db/gloda/modules/Gloda.sys.mjs
--- a/mailnews/db/gloda/modules/Gloda.sys.mjs
+++ b/mailnews/db/gloda/modules/Gloda.sys.mjs
@@ -966,17 +966,17 @@ export var Gloda = {
            *  the list of folders that match if gloda would index them.  This is
            *  unsuitable for producing a persistable constraint since it does not
            *  adapt for added/deleted folders.  However, it is sufficient for
            *  faceting.  Also, we don't persist constraints yet.
            *
            * TODO: The long-term solution is to move towards using arithmetic
            *     encoding on folder-id's like we use for MIME types and friends.
            */
-          Account(aAttrDef, aArguments) {
+          Account(aAttrDef, ...aArguments) {
             const folderValues = [];
             const seenRootFolders = {};
             for (let iArg = 0; iArg < aArguments.length; iArg++) {
               const givenFolder = aArguments[iArg];
               const givenMsgFolder = givenFolder.getXPCOMFolder(
                 givenFolder.kActivityFolderOnlyNoData
               );
               const rootFolder = givenMsgFolder.rootFolder;
diff --git a/mailnews/db/gloda/modules/NounMimetype.sys.mjs b/mailnews/db/gloda/modules/NounMimetype.sys.mjs
--- a/mailnews/db/gloda/modules/NounMimetype.sys.mjs
+++ b/mailnews/db/gloda/modules/NounMimetype.sys.mjs
@@ -528,17 +528,17 @@ export var MimeTypeNoun = {
     /**
      * Query for MIME type categories based on one or more MIME type objects
      *  passed in.  We want the range to span the entire block allocated to the
      *  category.
      *
      * @param {object} aAttrDef - The attribute that is using us. See Gloda.defineAttribute()
      * @param {any[]} aArguments - The actual arguments object.
      */
-    Category(aAttrDef, aArguments) {
+    Category(aAttrDef, ...aArguments) {
       const rangePairs = [];
       // If there are no arguments then we want to fall back to the 'in'
       //  constraint which matches on any attachment.
       if (!aArguments || aArguments.length == 0) {
         return this._inConstraintHelper(aAttrDef, []);
       }
 
       for (let iArg = 0; iArg < aArguments.length; iArg++) {

